### PR TITLE
[moe/zero] refactor low level optimizer

### PR DESCRIPTION
**Pull request by botbw**
# 📝 What does this PR do?

## TLDR: Refactor low-level ZeRO optimizer to better deal with param groups with different parallel strategies.

Basically, the original `LowLevelZeroOptimizer` was decoupled into a few classes (`LowLevelZeroOptimizer`, `LowLevelOptStrategyBase`, `LowLevelOptStrategy`, `MoeZeroStrategy`), where each `param_group` and `process_group` are bonded into one `LowLevelOptStrategyBase`.

`LowLevelOptStrategyBase` utilizes previous `XXXStore`s and manages the grad communication and synchronization in low level, and the children classes handle the grad manipulation  (between `working_params` and `master_params`) in tensor level.

In this way, different param_groups might use different zero and comm policies and do not interfere with each other, and they can have their own comm logic.

Other things we can do: refactor `XXXStore`s, previous `XXXStore`s manage all `param_groups` by `group_id`, which is a bit complex. however currently this logic is moved to `LowLevelOptStrategyBase` and `group_id == 0` is used by each store (search `DEFAULT_STORE_GROUP_ID`).

WARNING: Reviewing this is going to be painful.

### CODE REVIEW

File: tests/test_moe/test_moe_zero_fwd_bwd.py
Comment (botbw) [inline tests/test_moe/test_moe_zero_fwd_bwd.py:1]: Note: This test cannot run since we no longer have `MoeModel` on the `feature/moe` branch.

File: tests/test_moe/test_moe_zero_optim.py
Comment (botbw) [inline tests/test_moe/test_moe_zero_optim.py:1]: Note: This test cannot run since we no longer have `MoeModel` on the `feature/moe` branch.

File: tests/test_moe/test_moe_load_balance.py [hunk 1]
```diff
@@ -176,6 +176,7 @@ def run_dist(rank, world_size, port):
     run_hybrid_zero_optim_test(rank, world_size, stage=2)
 
 
+@pytest.mark.skip(reason="moe need to be refactored")
```
Comment (botbw) [inline tests/test_moe/test_moe_load_balance.py:179]: Note: This test cannot run since we no longer have `MoeModel` on the `feature/moe` branch.

**Review by Hz188 — APPROVED** [review]

**Review by Hz188 — APPROVED** [review]
